Sources/GRPC/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -342,3 +417,73 @@ extension HTTP1ToRawGRPCServerCodec: ChannelOutboundHandler {
     }
   }
 }
+
+private extension HTTP1ToRawGRPCServerCodec {
+  enum RequestEncodingValidation {
+    /// The compression is not supported. The RPC should fail with an appropriate status and include
+    /// our supported algorithms in the trailers.
+    case unsupported
+
+    /// Compression is supported.
+    case supported(CompressionAlgorithm)
+
+    /// Compression is supported but we did not disclose our support for it. We should continue but
+    /// also send the acceptable compression methods (including the encoding the client specified)
+    /// in the initial response metadata.
+    case supportedButNotDisclosed(CompressionAlgorithm)
+  }
+
+  /// Validates the value of the 'grpc-encoding' header against compression algorithms supported and
+  /// advertised by this peer.
+  ///
+  /// - Parameter requestEncoding: The value of the 'grpc-encoding' header.
+  func validate(requestEncoding: String) -> RequestEncodingValidation {
+    guard let algorithm = CompressionAlgorithm(rawValue: requestEncoding) else {
+      return .unsupported
+    }
+
+    if self.encoding.acceptableRequestCompression.contains(algorithm) {
+      return .supported(algorithm)
+    } else {
+      return .supportedButNotDisclosed(algorithm)
+    }
+  }
+
+  /// Makes a 'grpc-accept-encoding' header from the advertised encodings and an additional value
+  /// if one is specified.
+  func makeAcceptEncodingHeader(includeExtra extra: CompressionAlgorithm? = nil) -> String? {
+    switch (self.encoding.acceptableRequestCompression.isEmpty, extra) {
+    case (false, .some(let extra)):
+      return (self.encoding.acceptableRequestCompression + CollectionOfOne(extra)).map { $0.name }.joined(separator: ",")
+    case (false, .none):
+      return self.encoding.acceptableRequestCompression.map { $0.name }.joined(separator: ",")
+    case (true, .some(let extra)):
+      return extra.name
+    case (true, .none):
+      return nil
+    }
+  }
+
+  /// Selects an appropriate response encoding from the list of encodings sent to us by the client.
+  /// Returns `nil` if there were no appropriate algorithms, in which case the server will send
+  /// messages uncompressed.
+  func selectResponseEncoding(from acceptableEncoding: [Substring]) -> CompressionAlgorithm? {
+    guard let preference = self.encoding.preferredResponseCompression, preference != .identity else {
+      return nil
+    }
+
+    // What do we both support?
+    let mutuallySupported = acceptableEncoding.compactMap {
+      CompressionAlgorithm(rawValue: String($0))
+    }
+
+    // Is our preferred mechanism there?
+    if mutuallySupported.contains(preference) {
```

Comment:
This sounds like _we_ pick the preferred encoding; shouldn't we instead try returning the first supported encoding requested by the client (similar to the `Accept-Language` header)? (I'm not that familiar with the gRPC spec; let me know if I'm mistaken here.)

Reply:
The gRPC spec is _really_ unclear about compression, most of it relates to the API in libgrpc. I think we should probably do as you suggest, it's easier to understand and simplifies the configuration. Perhaps it would be better to have the configuration decide which compression algorithms are acceptable (defaulting to all). On the call context we could then give the user a choice to enable/disable compression per RPC (and per message on the `send` methods). Obviously if no compression was negotiated then enable/disable will be ignored. WDYT?

In fact there's probably no reason we can't do similar for the client, have a per RPC configuration instead of restricting it to the connection.

Reply:
Sounds good to me, but to be honest I believe most users won't ever need to touch this, neither on the per-connection nor the per-RPC level. I would default to turning everything on and allowing to turn it off (either per-connection or per-RPC, both is fine with me).